File: scala/phone-number/src/main/scala/PhoneNumber.scala
class PhoneNumber(val num: String) {
  def number(): Option[String] = {
    val numParsed = num.filter(_.isDigit).toString
    numParsed.length match {
      case l if l < 10 || l > 11 => return None
      case l if l == 10 => return Some(numParsed)
      case l if l == 11 && numParsed.take(1) == "1" =>
        return Some(numParsed.drop(1))
      case _ => None

    }
  }

  def areaCode(): Option[String] = {
    this.number match {
      case Some(n) => Some(n.take(3))
      case None => None
    }
  }

  def prettyPrint(): Option[String] = {
    this.number match {
      case Some(n) =>
        Some(s"(${n.take(3)}) ${n.drop(3).take(3)}-${n.drop(6).take(4)}")
      case None => None
    }
  }

}

File: javascript/phone-number/phone-number.js
const Immutable = require("Immutable")

class PhoneNumber {
    constructor(phone) {
        this.phone = phone
    }

    number() {
        const numbers = Immutable.Set.of(...
            "0123456789");
        let number = Immutable.List.of(...this.phone).filter(ch => numbers.has(ch));
        const badNumber = "0000000000";
        if (number.size < 10 || number.size > 12) return badNumber;
        if (number.size === 11) {
            if (number.first() === "1") {
                number = number.rest()
            } else {
                return badNumber;
            }
        }
        return number.join("");
    }

    areaCode() {
        return this.number().substr(0, 3);
    }

    toString() {
        let n = this.number();
        return `(${this.areaCode()}) ${n.substr(3, 3)}-${n.substr(6, 4)}`
    }
}

module.exports = PhoneNumber